### PR TITLE
feat: add levenshtein distance check for password validation

### LINKED ISSUE #184: Require Levenshtein distance between identifiers and password
**Is your feature request related to a problem? Please describe.**

It's currently possible to use a password during registration which is almost identical to the username (e.g. `foo@bar.com` and `foo@bar.com1`).

**Describe the solution you'd like**

This should be disallowed by the password policy. We can probably leverage the Levenshtein distance to ensure that the two strings are "different enough".

### PR DESCRIPTION
**Pull request by zepatrik**
## Related issue
closes #184 

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [ ] I have read the [security policy](../security/policy)
- [ ] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

File: selfservice/strategy/password/validator.go [hunk 1]
```diff
@@ -46,14 +48,17 @@ type DefaultPasswordValidator struct {
...
 	return &DefaultPasswordValidator{
-		c:                    http.DefaultClient,
-		maxBreachesThreshold: 0,
-		hashes:               map[string]int64{},
-		ignoreNetworkErrors:  true,
+		c:                         http.DefaultClient,
+		maxBreachesThreshold:      0,
+		hashes:                    map[string]int64{},
+		ignoreNetworkErrors:       true,
+		minIdentifierPasswordDist: 5,
```
Comment (aeneasr) [inline selfservice/strategy/password/validator.go]: Can we make this configurable in the config file, and also document this in ? :)

Comment (aeneasr) [reply selfservice/strategy/password/validator.go]: https://www.ory.sh/docs/kratos/self-service/strategies/username-email-password#security-and-defenses

File: selfservice/strategy/password/validator_test.go [hunk 2]
```diff
@@ -30,6 +50,12 @@ func TestDefaultPasswordValidationStrategy(t *testing.T) {
 		{pw: "l3f9toh1uaf81n21", id: "l3f9toh1uaf81n21", pass: false},
 		{pw: "l3f9toh1", pass: true},
 		{pw: "000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000", pass: true},
+		// simple permutation tests
+		{id: "hello@example.com", pw: "hello@example.com12345", pass: false},
+		{id: "hello@example.com", pw: "123hello@example.com123", pass: false},
+		{id: "hello@example.com", pw: "hello@exam", pass: false},
+		{id: "hello@example.com", pw: "HELLO@EXAMPLE.COM", pass: false},
```
Comment (aeneasr) [inline selfservice/strategy/password/validator_test.go]: Let's add more test cases:

```suggestion
		{id: "hello@example.com", pw: "HELLO@EXAMPLE.COM", pass: false},
		{id: "hello@example.com", pw: "HELLO@example.com", pass: false},
		{pw: "hello@example.com", id: "hello@exam", pass: false},
		{id: "hello@example.com", pw: "h3ll0@example", pass: false},
		{pw: "hello@example.com", id: "hello@exam", pass: false},
		{pw: "abcd", id: "9d3c8a1b", pass: true},
		{pw: "a", id: "kjOkla", pass: true},
```

File: selfservice/strategy/password/validator_test.go [hunk 3]
```diff
@@ -30,6 +50,18 @@ func TestDefaultPasswordValidationStrategy(t *testing.T) {
...
+		{id: "hello@example.com", pw: "HELLO@example.com", pass: false},
+		{pw: "hello@example.com", id: "hello@exam", pass: false},
+		{id: "hello@example.com", pw: "h3ll0@example", pass: false},
+		{pw: "hello@example.com", id: "hello@exam", pass: false},
+		{pw: "abcd", id: "9d3c8a1b", pass: true},
+		{pw: "a", id: "kjOkla", pass: true},
```
Comment (aeneasr) [inline selfservice/strategy/password/validator_test.go]:
```suggestion
		{id: "a", pw: "kjOkla", pass: true},
```

File: selfservice/strategy/password/validator_test.go [hunk 4]
```diff
@@ -30,6 +50,18 @@ func TestDefaultPasswordValidationStrategy(t *testing.T) {
...
+		// simple permutation tests
+		{id: "hello@example.com", pw: "hello@example.com12345", pass: false},
+		{id: "hello@example.com", pw: "123hello@example.com123", pass: false},
+		{id: "hello@example.com", pw: "hello@exam", pass: false},
+		{id: "hello@example.com", pw: "HELLO@EXAMPLE.COM", pass: false},
+		{id: "hello@example.com", pw: "HELLO@example.com", pass: false},
+		{pw: "hello@example.com", id: "hello@exam", pass: false},
+		{id: "hello@example.com", pw: "h3ll0@example", pass: false},
+		{pw: "hello@example.com", id: "hello@exam", pass: false},
+		{pw: "abcd", id: "9d3c8a1b", pass: true},
```
Comment (aeneasr) [inline selfservice/strategy/password/validator_test.go]:
```suggestion
		{id: "abcd", pw: "9d3c8a1b", pass: true},
```

File: selfservice/strategy/password/validator.go (hunk 1 above)
Comment (Kaishu-Sahu) [reply selfservice/strategy/password/validator.go]: I think, this thing has been implemented. Right?

`func (s *DefaultPasswordValidator) Validate(ctx context.Context, identifier, password string) error {
	passwordPolicyConfig := s.reg.Config(ctx).PasswordPolicyConfig()
	if len(password) < int(passwordPolicyConfig.MinPasswordLength) {`

Comment (zepatrik) [reply selfservice/strategy/password/validator.go]: You mean the configuration? Does not look like the values come from the config rn: https://github.com/ory/kratos/blob/10778fdd16a116b5dc8f4c2bdc96a895728d9aec/selfservice/strategy/password/validator.go#L155

**Pull request merged**